File: faunadb-scala/src/main/scala/com/faunadb/client/response/Codec.scala
package com.faunadb.client.response

import com.fasterxml.jackson.core.JsonParser
import com.fasterxml.jackson.databind._
import com.fasterxml.jackson.databind.node.ObjectNode

class ResponseNodeDeserializer extends JsonDeserializer[ResponseNode] {
  override def deserialize(jsonParser: JsonParser, ctxt: DeserializationContext): ResponseNode = {
    val json = jsonParser.getCodec.asInstanceOf[ObjectMapper]
    val tree = json.readTree(jsonParser).asInstanceOf[JsonNode]
    new ResponseNode(tree, json)
  }
}

class ResponseMapDeserializer extends JsonDeserializer[ResponseMap] {
  override def deserialize(jsonParser: JsonParser, ctxt: DeserializationContext): ResponseMap = {
    val json = jsonParser.getCodec.asInstanceOf[ObjectMapper]
    val tree = json.readTree(jsonParser).asInstanceOf[JsonNode]

    val innerTree = if(tree.has("@obj"))
      tree.get("@obj").asInstanceOf[ObjectNode]
    else
      tree

    val t = ctxt.getTypeFactory.constructMapLikeType(classOf[Map[_,_]], classOf[String], classOf[ResponseNode])
    new ResponseMap(json.convertValue(innerTree, t))
  }
}

File: faunadb-java/src/main/java/com/faunadb/client/response/Codec.java
package com.faunadb.client.response;

import com.fasterxml.jackson.core.JsonParser;
import com.fasterxml.jackson.databind.DeserializationContext;
import com.fasterxml.jackson.databind.JsonDeserializer;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.type.MapLikeType;
import com.google.common.collect.ImmutableMap;

import java.io.IOException;

class Codec {
  public static class ResponseNodeDeserializer extends JsonDeserializer<ResponseNode> {
    @Override
    public ResponseNode deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
      ObjectMapper json = (ObjectMapper) jsonParser.getCodec();
      JsonNode tree = json.readTree(jsonParser);
      return ResponseNode.create(tree, json);
    }
  }

  public static class ResponseMapDeserializer extends JsonDeserializer<ResponseMap> {
    @Override
    public ResponseMap deserialize(JsonParser jsonParser, DeserializationContext deserializationContext) throws IOException {
      ObjectMapper json = (ObjectMapper) jsonParser.getCodec();
      JsonNode tree = json.readTree(jsonParser);

      ObjectNode innerTree;
      if (tree.has("@obj")) {
        innerTree = (ObjectNode) tree.get("@obj");
      } else {
        innerTree = (ObjectNode) tree;
      }

      MapLikeType t = deserializationContext.getTypeFactory().constructMapLikeType(ImmutableMap.class, String.class, ResponseNode.class);
      return new ResponseMap(json.<ImmutableMap<String, ResponseNode>>convertValue(innerTree, t));
    }
  }
}
